libs/zzz/solver/src/calc.ts

```diff
@@ -13,22 +17,58 @@ import {
 import type { Stats } from '@genshin-optimizer/zzz/db'
 import type { DiscStats } from './common'
 
+export function passSetFilter(
+  discs: DiscStats[],
+  filter2: DiscSetKey[],
+  filter4: DiscSetKey[]
+): boolean {
+  const setCount: Partial<Record<DiscSetKey, number>> = {}
+  if (!filter4.length && !filter2.length) return true
+  for (const { setKey } of discs) setCount[setKey] = (setCount[setKey] || 0) + 1
+
+  if (filter4.length) {
+    let has4 = false
+    for (const key in setCount) {
+      const k = key as DiscSetKey
+      if (setCount[k]! >= 4 && filter4.includes(k)) {
+        has4 = true
+        break // can break out because there is only one 4p
+      }
+    }
+    if (!has4) return false
+  }
+  if (filter2.length) {
+    for (const key in setCount) {
+      const k = key as DiscSetKey
+      const val = setCount[k]!
+      if (val === 1) return false // Rainbow
+      if (val >= 2 && val < 4 && !filter4.includes(k)) return false
```

Comment:
logic looks wrong? why would we exclude a 2 or 3p based on filter4?

Reply:
good catch